dali/pipeline/data/buffer.h

```diff
@@ -289,34 +339,65 @@ class DLL_PUBLIC Buffer {
     set_type(TypeTable::GetTypeId<T>());
   }
 
-  inline void reserve(size_t new_num_bytes) {
-    if (new_num_bytes <= num_bytes_) return;
-
-    // re-allocating: get the device
-    if (std::is_same<Backend, GPUBackend>::value) {
-      CUDA_CALL(cudaGetDevice(&device_));
-    } else {
-      device_ = CPU_ONLY_DEVICE_ID;
-    }
+  inline void reserve(size_t new_num_bytes, AccessOrder order = {}) {
+    if (new_num_bytes <= num_bytes_)
+      return;
 
     DALI_ENFORCE(!shares_data_,
                  "Cannot reallocate Buffer if it is sharing data. "
                  "Clear the status by `Reset()` first.");
-    data_.reset();
+
+    if (data_ && device_ >= 0 && order.is_device() && order.device_id() != device_)
+      DALI_FAIL("Cannot reallocate a buffer on a different device!");
+
+    set_order(order);
+    free_storage();
+
+    if (device_ < 0) {
+      if (order.is_device() && order.device_id() >= 0) {
+        device_ = order.device_id();
```

Comment:
Doesn't this make the CPU buffer a GPU one?

Reply:
It's for pinned buffers - they may be accessed by GPUs and are subject to stream semantics.

Reply:
Understood.